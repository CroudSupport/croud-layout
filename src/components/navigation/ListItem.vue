<template>
  <li :class="activeClass">
    <a v-if="item.children && item.children.length" @click.prevent="toggleItem" :id="itemSlug" :class="linkClass" :href="item.uri">
      <i v-if="item.icon != ''" :class="item.icon + ' link-icon'"></i>
      <span><i :class="linkIcon"></i>
          {{item.label}}
      </span>
    </a>
    <a v-else :id="itemSlug" :class="item.class" :href="item.uri" :target="item.target">
      <i v-if="item.icon != ''" :class="item.icon + ' link-icon'"></i>
      <span v-html="item.label"></span>
    </a>
    <div v-if="item.children && localItem.expanded" class="children">
        <ul :data-depth="depth+1" class="child-menu" >
        <list-item :item="child" :depth="depth+1" v-for="child in item.children" :key="child.uri" :expanded="expanded"></list-item>
        </ul>
    </div>
  </li>
</template>

<script>
    export default {
        name: 'ListItem',

        props: {
            item: {
                type: Object,
                required: true,
            },
            depth: {
                type: Number,
                default: 1,
            },
            expanded: {
                type: Boolean,
                required: true,
            },
        },

        data() {
            return {
                popoverTarget: null,
                localItem: this.item,
            }
        },

        computed: {
            activeClass() {
                if (!this.item.uri || this.item.uri.length < 1) return ''

                const uri = window.location.pathname.substr(1)
                return uri === this.item.uri.substring(1) ? 'active' : ''
            },

            linkIcon() {
                return this.localItem.expanded ? 'caret down icon expandable' : 'caret right icon expandable'
            },

            itemSlug() {
                return `menuitem-${this.item.label.toLowerCase().replace(/[^a-z0-9+]+/gi, '')}`
            },

            linkClass() {
                const itemClass = this.item.class || ''
                return this.localItem.expanded ? `${itemClass} sub-menu opened` : `${itemClass} sub-menu closed`
            },
        },

        methods: {
            toggleItem() {
                this.$emit('update:expanded', true)
                this.localItem.expanded = !this.localItem.expanded
            },
        },
    }
</script>
