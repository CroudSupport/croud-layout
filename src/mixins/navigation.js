import { checkPermissionHelper } from './security'

export default {
    computed: {
        filteredItems() {
            const recur = (item) => {
                if (item.children) {
                    item.children = item.children.filter(recur)
                }

                return !item.permission || checkPermissionHelper(this.permissions, item.permission)
            }

            return this.items.filter(recur)
        },

        baseItems() {
            const global = [
                {
                    label: 'Dashboard',
                    uri: '/',
                    icon: 'dashboard icon',
                },
                {
                    label: 'Bearhug',
                    uri: '/bearhug/index/standalone',
                    icon: 'paw icon',
                    permission: 'menu.dashboard.read',
                },
                {
                    label: 'Croud Academy',
                    uri: '/academy/',
                    icon: 'university icon',
                    expanded: false,
                    permission: 'menu.academy.read',
                    children: [
                        {
                            label: 'My Courses',
                            uri: '/academy/course/owned/',
                            permission: 'menu.academy.read',
                        },
                        {
                            label: 'Available Courses',
                            uri: '/academy/course/available/',
                            permission: 'menu.academy.read',
                        },
                        {
                            label: 'Review Submissions',
                            uri: '/academy/assessment/',
                            permission: 'academy.submission.manage',
                        },
                        {
                            label: 'Courses',
                            uri: '/academy/course/',
                            permission: 'academy.course.manage',
                        },
                        {
                            label: 'Modules',
                            uri: '/academy/modules/',
                            permission: 'academy.modules.manage',
                        },
                        {
                            label: 'Test',
                            uri: '/survey/',
                            permission: 'academy.survey.manage',
                        },
                        {
                            label: 'Qualifications',
                            uri: '/academy/qualification/',
                            permission: 'academy.qualification.manage',
                        },
                        {
                            label: 'Skillsets',
                            uri: '/academy/skillset/',
                            permission: 'academy.skillset.manage',
                        },
                    ],
                },
                {
                    label: 'User Management',
                    uri: '/users/manager',
                    icon: 'users icon',
                    permission: 'menu.manageUsers.read',
                },
                {
                    label: 'Knowledge Base',
                    uri: '/knowledge-base/',
                    icon: 'help circle icon',
                    permission: 'menu.knowledgebase.read',
                },
                {
                    label: 'Finance &amp; Invoicing',
                    uri: '/admin/finance',
                    icon: 'money icon',
                    permission: 'menu.finance.read',
                },
                {
                    label: 'Reporting',
                    uri: '/reporting/index/',
                    icon: 'bar chart icon',
                    permission: 'reports.reports.report',
                },
                {
                    label: 'System Settings',
                    uri: '/default/system/index',
                    icon: 'settings icon',
                    permission: 'menu.system.read',
                },
                {
                    label: 'My Details',
                    uri: '/user/account',
                    icon: 'user icon',
                },
            ]

            if (this.user.system) {
                return [
                    ...global,
                    {
                        label: 'Datorama',
                        uri: 'https://performance.croudcontrol.com',
                        icon: 'datorama icon',
                        class: 'external-link datorama-link',
                        target: '_blank',
                    },

                    {
                        label: 'Schoox',
                        uri: 'https://www.schoox.com/login.php',
                        icon: 'schoox icon',
                        class: 'external-link schoox-link',
                        target: '_blank',
                    },

                    {
                        label: 'Croud HR',
                        uri: 'https://croud.peoplehr.net',
                        icon: 'peoplehr icon icon',
                        class: 'external-link peoplehr-link',
                        target: '_blank',
                    },
                ]
            }

            if (this.user.onboarding) {
                return [
                    {
                        label: 'My Academy Courses',
                        uri: '/academy/course/owned/',

                        permission: 'menu.academy.read',
                    },
                    {
                        label: 'Available Academy Courses',
                        uri: '/academy/course/owned/',

                        permission: 'menu.academy.read',
                    },
                    {
                        label: 'My Details',
                        uri: '/user/account',
                        icon: 'user icon',

                    },
                ]
            }

            return [
                ...global,
                {
                    label: 'Help Centre',
                    uri: 'https://croud.freshdesk.com/support/solutions',
                    icon: 'help circle outline icon',

                },
            ]
        },

        items() {
            return this.baseItems
        },
    },
}
