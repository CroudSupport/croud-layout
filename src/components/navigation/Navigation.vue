<template>
  <div class="contained scrollable navigation-block">
    <div class="footer">
      <strong class="expand-button" @click.prevent="expanded = !expanded">
          <i :class="{left: expanded, right: !expanded}" class="arrow circle double icon"></i>
      </strong>
    </div>
    <ul data-depth="1">
      <list-item
      v-for="item in filteredItems"
      :item="item"
      :expanded.sync="expanded"
      :key="item.uri"
      />
    </ul>
  </div>
</template>

<script>
    import { mapGetters } from 'vuex'
    import mixin from '../../mixins/navigation'

    import ListItem from './ListItem'

    export default {
        mixins: [mixin],

        components: {
            ListItem,
        },

        computed: {
            expanded: {
                get() {
                    return this.$store.getters['universal/expandedNav']
                },
                set(value) {
                    this.$store.commit('universal/EXPAND_NAV', value)
                },
            },

            ...mapGetters({
                permissions: 'universal/effectivePermissions',
                user: 'universal/user',
            }),
        },
    }
</script>
