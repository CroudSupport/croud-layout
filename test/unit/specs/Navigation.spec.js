import Vue from 'vue'
import navigation from '../../../src/mixins/navigation'

const vm = new Vue({
    data() {
        return {
            user: {
                system: false,
                onboarding: false,
            },
            permissions: [
                'menu.dashboard.read',
                'menu.academy.read',
                'academy.submission.manage',
                'menu.manageUsers.read',
            ],
        }
    },
    computed: {
        items() {
            return this.baseItems
        },
    },
    render: h => h(),
    mixins: [navigation],
}).$mount()

describe('Navigation mixin', () => {
    describe('Base items array', () => {
        afterAll((next) => {
            vm.user = {
                system: true,
                onboarding: false,
            }

            vm.$nextTick(next)
        })
        it('should produce base SC menu', () => {
            expect(typeof vm.items).toEqual('object')
            expect(vm.items).toMatchSnapshot()
        })

        it('should produce base croudie menu', () => {
            vm.user.system = false
            vm.$nextTick(() => {
                expect(typeof vm.items).toEqual('object')
                expect(vm.items).toMatchSnapshot()
            })
        })

        it('should produce base onboarding croudie menu', () => {
            vm.user = {
                system: false,
                onboarding: true,
            }
            vm.$nextTick(() => {
                expect(typeof vm.items).toEqual('object')
                expect(vm.items).toMatchSnapshot()
            })
        })
    })

    describe('Filtered item array', () => {
        afterAll((next) => {
            vm.user = {
                system: true,
                onboarding: false,
            }

            vm.$nextTick(next)
        })

        describe('SC menu', () => {
            it('should produce menu', () => {
                vm.$nextTick(() => {
                    expect(typeof vm.filteredItems).toEqual('object')
                    expect(vm.filteredItems).toMatchSnapshot()
                })
            })

            it('should build children menu', () => {
                vm.$nextTick(() => {
                    expect(Array.isArray(vm.filteredItems.filter(i => i.children)[0].children)).toEqual(true)
                    expect(vm.filteredItems.filter(i => i.children)[0].children.length).toEqual(3)
                })
            })
        })

        describe('Croudie menu', () => {
            it('should produce menu', () => {
                vm.user.system = false
                vm.$nextTick(() => {
                    expect(vm.filteredItems).toMatchSnapshot()
                })
            })
        })

        describe('Onboarding Croudie menu', () => {
            it('should produce menu', () => {
                vm.user = {
                    system: false,
                    onboarding: true,
                }
                vm.$nextTick(() => {
                    expect(vm.filteredItems).toMatchSnapshot()
                })
            })
        })
    })
})
