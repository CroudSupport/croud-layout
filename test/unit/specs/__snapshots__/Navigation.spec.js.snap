// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation mixin Base items array should produce base SC menu 1`] = `
Array [
  Object {
    "icon": "dashboard icon",
    "label": "Dashboard",
    "uri": "/",
  },
  Object {
    "icon": "paw icon",
    "label": "Bearhug",
    "permission": "menu.dashboard.read",
    "uri": "/bearhug/index/standalone",
  },
  Object {
    "children": Array [
      Object {
        "label": "My Courses",
        "permission": "menu.academy.read",
        "uri": "/academy/course/owned/",
      },
      Object {
        "label": "Available Courses",
        "permission": "menu.academy.read",
        "uri": "/academy/course/available/",
      },
      Object {
        "label": "Review Submissions",
        "permission": "academy.submission.manage",
        "uri": "/academy/assessment/",
      },
      Object {
        "label": "Courses",
        "permission": "academy.course.manage",
        "uri": "/academy/course/",
      },
      Object {
        "label": "Modules",
        "permission": "academy.modules.manage",
        "uri": "/academy/modules/",
      },
      Object {
        "label": "Test",
        "permission": "academy.survey.manage",
        "uri": "/survey/",
      },
      Object {
        "label": "Qualifications",
        "permission": "academy.qualification.manage",
        "uri": "/academy/qualification/",
      },
      Object {
        "label": "Skillsets",
        "permission": "academy.skillset.manage",
        "uri": "/academy/skillset/",
      },
    ],
    "expanded": false,
    "icon": "university icon",
    "label": "Croud Academy",
    "permission": "menu.academy.read",
    "uri": "/academy/",
  },
  Object {
    "icon": "users icon",
    "label": "User Management",
    "permission": "menu.manageUsers.read",
    "uri": "/users/manager",
  },
  Object {
    "icon": "help circle icon",
    "label": "Knowledge Base",
    "permission": "menu.knowledgebase.read",
    "uri": "/knowledge-base/",
  },
  Object {
    "icon": "money icon",
    "label": "Finance &amp; Invoicing",
    "permission": "menu.finance.read",
    "uri": "/admin/finance",
  },
  Object {
    "icon": "bar chart icon",
    "label": "Reporting",
    "permission": "reports.reports.report",
    "uri": "/reporting/index/",
  },
  Object {
    "icon": "settings icon",
    "label": "System Settings",
    "permission": "menu.system.read",
    "uri": "/default/system/index",
  },
  Object {
    "icon": "user icon",
    "label": "My Details",
    "uri": "/user/account",
  },
  Object {
    "icon": "help circle outline icon",
    "label": "Help Centre",
    "uri": "https://croud.freshdesk.com/support/solutions",
  },
]
`;

exports[`Navigation mixin Base items array should produce base croudie menu 1`] = `
Array [
  Object {
    "icon": "dashboard icon",
    "label": "Dashboard",
    "uri": "/",
  },
  Object {
    "icon": "paw icon",
    "label": "Bearhug",
    "permission": "menu.dashboard.read",
    "uri": "/bearhug/index/standalone",
  },
  Object {
    "children": Array [
      Object {
        "label": "My Courses",
        "permission": "menu.academy.read",
        "uri": "/academy/course/owned/",
      },
      Object {
        "label": "Available Courses",
        "permission": "menu.academy.read",
        "uri": "/academy/course/available/",
      },
      Object {
        "label": "Review Submissions",
        "permission": "academy.submission.manage",
        "uri": "/academy/assessment/",
      },
      Object {
        "label": "Courses",
        "permission": "academy.course.manage",
        "uri": "/academy/course/",
      },
      Object {
        "label": "Modules",
        "permission": "academy.modules.manage",
        "uri": "/academy/modules/",
      },
      Object {
        "label": "Test",
        "permission": "academy.survey.manage",
        "uri": "/survey/",
      },
      Object {
        "label": "Qualifications",
        "permission": "academy.qualification.manage",
        "uri": "/academy/qualification/",
      },
      Object {
        "label": "Skillsets",
        "permission": "academy.skillset.manage",
        "uri": "/academy/skillset/",
      },
    ],
    "expanded": false,
    "icon": "university icon",
    "label": "Croud Academy",
    "permission": "menu.academy.read",
    "uri": "/academy/",
  },
  Object {
    "icon": "users icon",
    "label": "User Management",
    "permission": "menu.manageUsers.read",
    "uri": "/users/manager",
  },
  Object {
    "icon": "help circle icon",
    "label": "Knowledge Base",
    "permission": "menu.knowledgebase.read",
    "uri": "/knowledge-base/",
  },
  Object {
    "icon": "money icon",
    "label": "Finance &amp; Invoicing",
    "permission": "menu.finance.read",
    "uri": "/admin/finance",
  },
  Object {
    "icon": "bar chart icon",
    "label": "Reporting",
    "permission": "reports.reports.report",
    "uri": "/reporting/index/",
  },
  Object {
    "icon": "settings icon",
    "label": "System Settings",
    "permission": "menu.system.read",
    "uri": "/default/system/index",
  },
  Object {
    "icon": "user icon",
    "label": "My Details",
    "uri": "/user/account",
  },
  Object {
    "icon": "help circle outline icon",
    "label": "Help Centre",
    "uri": "https://croud.freshdesk.com/support/solutions",
  },
]
`;

exports[`Navigation mixin Base items array should produce base onboarding croudie menu 1`] = `
Array [
  Object {
    "label": "My Academy Courses",
    "permission": "menu.academy.read",
    "uri": "/academy/course/owned/",
  },
  Object {
    "label": "Available Academy Courses",
    "permission": "menu.academy.read",
    "uri": "/academy/course/available/",
  },
  Object {
    "icon": "user icon",
    "label": "My Details",
    "uri": "/user/account",
  },
]
`;

exports[`Navigation mixin Filtered item array Croudie menu should produce menu 1`] = `
Array [
  Object {
    "icon": "dashboard icon",
    "label": "Dashboard",
    "uri": "/",
  },
  Object {
    "icon": "paw icon",
    "label": "Bearhug",
    "permission": "menu.dashboard.read",
    "uri": "/bearhug/index/standalone",
  },
  Object {
    "children": Array [
      Object {
        "label": "My Courses",
        "permission": "menu.academy.read",
        "uri": "/academy/course/owned/",
      },
      Object {
        "label": "Available Courses",
        "permission": "menu.academy.read",
        "uri": "/academy/course/available/",
      },
      Object {
        "label": "Review Submissions",
        "permission": "academy.submission.manage",
        "uri": "/academy/assessment/",
      },
    ],
    "expanded": false,
    "icon": "university icon",
    "label": "Croud Academy",
    "permission": "menu.academy.read",
    "uri": "/academy/",
  },
  Object {
    "icon": "users icon",
    "label": "User Management",
    "permission": "menu.manageUsers.read",
    "uri": "/users/manager",
  },
  Object {
    "icon": "user icon",
    "label": "My Details",
    "uri": "/user/account",
  },
  Object {
    "icon": "help circle outline icon",
    "label": "Help Centre",
    "uri": "https://croud.freshdesk.com/support/solutions",
  },
]
`;

exports[`Navigation mixin Filtered item array Onboarding Croudie menu should produce menu 1`] = `
Array [
  Object {
    "label": "My Academy Courses",
    "permission": "menu.academy.read",
    "uri": "/academy/course/owned/",
  },
  Object {
    "label": "Available Academy Courses",
    "permission": "menu.academy.read",
    "uri": "/academy/course/available/",
  },
  Object {
    "icon": "user icon",
    "label": "My Details",
    "uri": "/user/account",
  },
]
`;

exports[`Navigation mixin Filtered item array SC menu should produce menu 1`] = `
Array [
  Object {
    "icon": "dashboard icon",
    "label": "Dashboard",
    "uri": "/",
  },
  Object {
    "icon": "paw icon",
    "label": "Bearhug",
    "permission": "menu.dashboard.read",
    "uri": "/bearhug/index/standalone",
  },
  Object {
    "children": Array [
      Object {
        "label": "My Courses",
        "permission": "menu.academy.read",
        "uri": "/academy/course/owned/",
      },
      Object {
        "label": "Available Courses",
        "permission": "menu.academy.read",
        "uri": "/academy/course/available/",
      },
      Object {
        "label": "Review Submissions",
        "permission": "academy.submission.manage",
        "uri": "/academy/assessment/",
      },
    ],
    "expanded": false,
    "icon": "university icon",
    "label": "Croud Academy",
    "permission": "menu.academy.read",
    "uri": "/academy/",
  },
  Object {
    "icon": "users icon",
    "label": "User Management",
    "permission": "menu.manageUsers.read",
    "uri": "/users/manager",
  },
  Object {
    "icon": "user icon",
    "label": "My Details",
    "uri": "/user/account",
  },
  Object {
    "class": "external-link datorama-link",
    "icon": "datorama icon",
    "label": "Datorama",
    "target": "_blank",
    "uri": "https://performance.croudcontrol.com",
  },
  Object {
    "class": "external-link schoox-link",
    "icon": "schoox icon",
    "label": "Schoox",
    "target": "_blank",
    "uri": "https://www.schoox.com/login.php",
  },
  Object {
    "class": "external-link peoplehr-link",
    "icon": "peoplehr icon icon",
    "label": "Croud HR",
    "target": "_blank",
    "uri": "https://croud.peoplehr.net",
  },
]
`;
